Enable nxp 4 channel dma driver

ZhaoxiangJin opened the pull request:
1. enabled nxp,4ch-dma driver
2. enabled nxp,4ch-dma driver on frdm_mcxc444
3. tested nxp,4ch-dma driver with dma/loop_transfer test case
test result:
```
*** Booting Zephyr OS build v4.2.0-6156-g17fe21076a3b ***
Running TESTSUITE dma_m2m_loop
===================================================================
START - test_tst_dma0_m2m_loop
DMA memory to memory transfer started
Preparing DMA Controller: dma-controller@40008000
Starting the transfer on channel 0 and waiting for 1 second
Each RX buffer should contain the full TX buffer string.
RX data Loop 0
RX data Loop 1
RX data Loop 2
RX data Loop 3
Finished DMA: dma-controller@40008000
 PASS - test_tst_dma0_m2m_loop in 0.279 seconds
===================================================================
START - test_tst_dma0_m2m_loop_repeated_start_stop
DMA memory to memory transfer started
Preparing DMA Controller
Starting the transfer on channel 0 and waiting for 1 second
Each RX buffer should contain the full TX buffer string.
RX data Loop 0
RX data Loop 1
RX data Loop 2
RX data Loop 3
Finished: DMA
 PASS - test_tst_dma0_m2m_loop_repeated_start_stop in 0.275 seconds
===================================================================
START - test_tst_dma0_m2m_loop_suspend_resume
DMA memory to memory transfer started
Preparing DMA Controller: dma-controller@40008000
Starting the transfer on channel 0 and waiting for 1 second
suspend not supported
 SKIP - test_tst_dma0_m2m_loop_suspend_resume in 0.017 seconds
===================================================================
TESTSUITE dma_m2m_loop succeeded

------ TESTSUITE SUMMARY START ------

SUITE PASS - 100.00% [dma_m2m_loop]: pass = 2, fail = 0, skip = 1, total = 3 duration = 0.571 seconds
 - PASS - [dma_m2m_loop.test_tst_dma0_m2m_loop] duration = 0.279 seconds
 - PASS - [dma_m2m_loop.test_tst_dma0_m2m_loop_repeated_start_stop] duration = 0.275 seconds
 - SKIP - [dma_m2m_loop.test_tst_dma0_m2m_loop_suspend_resume] duration = 0.017 seconds

------ TESTSUITE SUMMARY END ------

===================================================================
PROJECT EXECUTION SUCCESSFUL
```

Please access https://www.nxp.com/webapp/sps/download/preDownload.jsp?render=true to download MCXC44x Sub-Family Reference Manual for more details. DMAMUX can be found in Chapter 20 and DMA in Chapter 21.

decsny reviewed (dismissed):
the naming is confusing, calling it "nxp dma" when it is one of the most uncommon dma on nxp parts. I will defer to @DerekSnell to give naming feedback.

ZhaoxiangJin commented:
> > the naming is confusing, calling it "nxp dma" when it is one of the most uncommon dma on nxp parts. I will defer to @DerekSnell to give naming feedback.
> 
> Hi @decsny and @ZhaoxiangJin ,
> I agree with Declan and I do not like calling this just "nxp dma".  I suggest we call this LDMA to better differentiate.
> Thank you

Hello @DerekSnell , thanks for your suggestion, could you please explain  what 'L' stands for?

Holt-Sun commented on drivers/clock_control/clock_control_mcux_sim.c:
```diff
@@ -37,6 +37,16 @@ static int mcux_sim_on(const struct device *dev,
+
+	if ((uint32_t)sub_system == KINETIS_SIM_DMAMUX_CLK) {
+		clock_ip_name = kCLOCK_Dmamux0;
+	}
```
Risk of enabling wrong clock if an unsupported subsystem is passed after a previous match. Recommend explicit else return -ENOTSUP.

Holt-Sun commented on drivers/dma/dma_nxp_4ch_dma.c:
```diff
@@ -0,0 +1,474 @@
+	case 4:
+		return 0U; /* 32 bits */
+	default:
+		return 0U; /* default to 32 bits */
```
If others are unsupported, should LOG_WRN or LOG_ERR.

Holt-Sun commented on drivers/dma/dma_nxp_4ch_dma.c:
```diff
@@ -0,0 +1,474 @@
+	struct nxp_dma_data *data = dev->data;
+	DMA_Type *dma = cfg->dma_base;
+	struct nxp_dma_chan_data *chan_data = &data->chan[channel];
+
```
Better to add NULL pointer check for status.

```suggestion
    if (status == NULL) {
        return -EINVAL;
    }
```

mcuxted commented on drivers/dma/dma_nxp_4ch_dma.c:
```diff
@@ -0,0 +1,472 @@
+	chan_data->busy = true;
+
+	/* Start: either allow HW request or issue software start */
+	if ((dma->DMA[channel].DCR & DMA_DCR_ERQ_MASK) == 0U) {
```
It appears that the HW request has not been enabled. Can we determine which method to use for initiating DMA based on the transfer type(M2M or P2M)?

ZhaoxiangJin committed:
drivers: sim: Enable dma clock through sim driver

Enable dma and dmamux clock through sim driver

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
drivers: dma: nxp: Add support for NXP 4 channel DMA driver

This commit introduces a new DMA driver for NXP platforms,
specifically supporting the MCXC/kinetis series.

Please access https://www.nxp.com/webapp/sps/download/preDownload.jsp?render=true
to download MCXC44x Sub-Family Reference Manual for more details.
DMAMUX can be found in Chapter 20 and DMA in Chapter 21.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
boards: Enable DMA on frdm_mcxc444

Enable DMA on frdm_mcxc444

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
tests: dma: Enable loop_transfer for nxp,4ch-dma

Enable loop_transfer for nxp,4ch-dma

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin commented on drivers/dma/dma_nxp_4ch_dma.c:
```diff
@@ -0,0 +1,472 @@
+	}
+
+	/* P2P and P2M use HW request (ERQ = 1, CS = 1)
+	 * M2M and M2P use SW request (START = 1, ERQ = 0, CS = 0)
```
sorry for delayed update. I thought about it, and we shouldn't map `transfer type` to `trigger type`. Based on the hardware design, what we know is that hardware trigger definitely requires configuring dmamux. So we should enable hardware trigger when dmamux exists and slots are not 0, and use software trigger in all other cases. Code updated.

ZhaoxiangJin commented on tests/drivers/dma/loop_transfer/boards/frdm_mcxc444.conf:
```diff
@@ -0,0 +1,4 @@
+# Copyright 2025 NXP
+# SPDX-License-Identifier: Apache-2.0
+
+CONFIG_DMA_LOOP_TRANSFER_SIZE=1024
```
mcxc444 only have [32KB SRAM](https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/nxp/frdm_mcxc444/doc/index.rst), default CONFIG_DMA_LOOP_TRANSFER_SIZE is [8192](https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/drivers/dma/loop_transfer/Kconfig#L23), in app, this requires 8192 + 4 * 8192 bytes = 40KB , this exceeds the hardware capability.

teburd commented on drivers/dma/Kconfig.nxp_4ch_dma:
```diff
@@ -0,0 +1,7 @@
+# Copyright 2025
```
```suggestion
# Copyright 2025 NXP
```

teburd commented on drivers/dma/dma_nxp_4ch_dma.c:
```diff
@@ -0,0 +1,469 @@
+	clock_control_subsys_t dmamux_clk_subsys;
+#endif
+	/* Pointer to per-instance channel allocation bitmap */
+	atomic_t *channels_atomic;
```
Delete this and initialize dma_context with the DT property instead

teburd commented on drivers/dma/dma_nxp_4ch_dma.c:
```diff
@@ -0,0 +1,469 @@
+	DMA_Type *dma_base;
+	DMAMUX_Type *dmamux_base;
+	uint8_t num_channels;
+	uint8_t dma_requests;
```
looks like an unused member? delete

teburd commented on drivers/dma/dma_nxp_4ch_dma.c:
```diff
@@ -0,0 +1,469 @@
+	uint8_t num_channels;
+	uint8_t dma_requests;
+	void (*irq_config_func)(const struct device *dev);
+#if IS_ENABLED(CONFIG_CLOCK_CONTROL)
```
Will this driver work without clock control? I'm highly skeptical from what I'm seeing here. If you *need* clock control, then use Kconfig to ensure its selected when this driver is being built rather than ifdef'ing. If you can work without it I'd be curious to see how, are the clocks setup at soc or board init time?
